[privatebin] add guide for privatebin (#853)
Uberspace/lab: source/guide_privatebin.rst

PrivateBin
##########

.. tag_list::

PrivateBin is a minimalist, open source online pastebin where the server has zero knowledge of pasted data.

Data is encrypted and decrypted in the browser using 256bit AES in Galois Counter mode.

----

.. note:: For this guide you should be familiar with the basic concepts of

  * PHP_
  * :manual:`domains <web-domains>`

License
=======

PrivateBin consists of PHP and JS code which was originally written by Sébastien Sauvage in 2012 and falls unter the Zlib/libpng license.
All relevant legal information can be found in the Github_ repository of the project.

Prerequisites
=============

We're using :manual:`PHP <lang-php>` in the stable version 7.4:

::

 [isabell@stardust ~]$ uberspace tools version show node
 Using 'PHP' version: '7.4'
 [isabell@stardust ~]$

The domain you want to use must be set up:

.. include:: includes/web-domain-list.rst

Installation
============

Download the source
-------------------

Check Github_ for the `latest release`_ and copy the download link to the .tar.gz file.
Then ``cd`` to your :manual:`DocumentRoot <web-documentroot>` and use ``wget`` to download it. Replace the URL with the one you just copied.

> In the example the version is 1.2.4 which does not exist.

Reviewer: We had fake versions for the longest time, but it just ended up causing more harm than joy. We've updated the style guide since to say

> If possible use a download URL that points to the latest version (e.g. latest.zip on some platforms). If such an URL is not available, use the newest version instead.

Please update the guide to point to version 1.3.4.

.. code-block:: console
 :emphasize-lines: 2

>  [isabell@stardust ~]$ cd /var/www/virtual/$USER/

Reviewer: This seem a bit convoluted. First we `cd` into `html/..` to then just extract into `html` using `-C html/`. What do you think about changing this to `cd /var/www/virtual/$USER/html` (or rather `cd ~/html`, which is much shorter) and skipping the `-C`?

Please make sure to also update the prompts for the following lines to say `html]$` instead of `isabell]$`.

 [isabell@stardust isabell]$ wget https://github.com/PrivateBin/PrivateBin/archive/1.2.4.tar.gz
 […]
 Saving to: ‘1.2.4.tar.gz’

 100%[=================================================>] 3,172,029   3.45MB/s   in 0.9s

 2020-11-22 16:27:44 (8.32 MB/s) - ‘1.2.4.tar.gz’ saved [523648]

>  [isabell@stardust ~]$

Reviewer: Since there isn't another `cd` here, the prompt stays the same and doesn't revert back to `~`.

Author: The style-guide did confuse me here as it says "Always end your console code blocks with [isabell@stardust ~]$."  :-)

Untar the archive to the ``html`` folder and then delete it. Replace the version in the file name with the one you downloaded.

.. code-block:: console
 :emphasize-lines: 1,2

 [isabell@stardust isabell]$ tar -xzf 1.2.4.tar.gz -C html/ --strip-components=1
 [isabell@stardust isabell]$ rm 1.2.4.tar.gz
 [isabell@stardust ~]$

Activate the .htaccess file
-----------------------------
PrivateBin provides a .htaccess file, which blocks some known robots and link-scanning bots. Activate it by renaming it from ``.htaccess.disabled`` to ``.htaccess``.

.. code-block:: console

>  [isabell@stardust ~]$ cd /var/www/virtual/$USER/html

Reviewer: If you accept my previous suggestion, we're already in `html` and you can skip this line entirely.
```suggestion
```

 [isabell@stardust html]$ mv .htaccess.disabled .htaccess

>  [isabell@stardust ~]$

Reviewer: prompt stays the same, like above.

Moving files outside of DocumentRoot
------------------------------------
It is recommended_ to move the configuration, data files, templates and PHP libraries outside of your document root.
To do that, create a Folder  ``privatebin-data`` in ``/var/www/virtual/isabell/`` and move the folders to the new location.
If not already there, go to the ``html`` directory before running ``mv``.

.. code-block:: console

>  [isabell@stardust html]$ mkdir /var/www/virtual/$USER/privatebin-data

Reviewer: What do you think about using a relative path to keep the command short?
```suggestion
 [isabell@stardust html]$ mkdir ../privatebin-data
```

>  [isabell@stardust html]$ cd /var/www/virtual/$USER/html

Reviewer: the prompt and prior steps indicate that we're already in `html`. No need to change the directory here.
```suggestion
```

>  [isabell@stardust html]$ mv -t /var/www/virtual/$USER/privatebin-data cfg/ lib/ tpl/ vendor/

Reviewer: same as above, relative?
```suggestion
 [isabell@stardust html]$ mv -t ../privatebin-data cfg/ lib/ tpl/ vendor/
```

>  [isabell@stardust ~]$

Reviewer: prompt stays the same :) just adding a note to each so I keep track of them ;)

Changing index.php
------------------

Now edit ``/var/www/virtual/$USER/html/index.php``  to inform PrivateBin about to the new location of the folders. 

.. warning:: Replace ``<username>`` with your username!

.. code-block:: php
 :emphasize-lines: 3

 [...]
 // change this, if your php files and data is outside of your webservers document root

>  define('PATH', '/var/www/virtual/<username>/privatebin-data/');

Reviewer: Maybe this also works with a relative path like above? This would remove the additional step of replacing `<username>`.

Author: it works 👍 Nice!

 [...]

> Setting permissions

Reviewer: The permissions look fine without this step. Did you encounter permission problems while creating the guide?

Author: There were no problems but the privatebin-data folder has 775 instead of 755 and the files 664 instead of 644. Is that a problem? I guess not, so I removed the chapter.

-------------------

Let's make sure the files and folder have the correct :manual_anchor:`permissions <web-documentroot#permissions>`.
Use ``find`` in combination with ``chmod`` to change them.

.. code-block:: console

 [isabell@stardust ~]$ cd /var/www/virtual/$USER
 [isabell@stardust isabell]$ find privatebin-data -type d -exec chmod 755 {} \;
 [isabell@stardust isabell]$ find privatebin-data -type f -exec chmod 644 {} \;
 [isabell@stardust ~]$

Configuration
=============

Configure your PrivateBin Instance
----------------------------------

> .. note:: You are not forced to change any of the default settings as they are mostly secure.

Reviewer: "forced" is a bit of a strong word :sweat_smile: 
```suggestion
.. note:: You don't need to change any of the default settings as they are mostly secure.
```

Author: I agree :-)
